psl-core/src/main/java/org/linqs/psl/model/rule/AbstractRule.java

```diff
@@ -17,16 +17,29 @@
  */
 package org.linqs.psl.model.rule;
 
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 /**
  * Base class for all (first order, i.e., not ground) rules.
  */
 public abstract class AbstractRule implements Rule {
     protected final String name;
 
+    private static Map<Integer, AbstractRule> Rules = new HashMap<Integer, AbstractRule>();
+
+    public static AbstractRule getRule(int hashcode) {
+        return Rules.get(hashcode);
+    }
+
+    public static void addRule(AbstractRule rule) {
+        Rules.putIfAbsent(System.identityHashCode(rule), rule);
```

Comment:
We are still supporting Java 7, so stick with `put`.
In this specific case, we are safe since `rule` is actually `this` in the constructor.
(This is part of the "coupling" I mentioned in another comment.)